dist/clouduploader.rb

```diff
@@ -4,99 +4,181 @@
+      image_name = calculate_image_name(image_filename) if image_name.empty?
+      uncompressed_file = uncompress(backend_image_file)
+      remote_file_name = File.basename(uncompressed_file)
+      application_id, application_key = decrypt([job_data['application_id'], job_data['application_key']])
 
-if platform == 'ec2'
-  upload_image_to_ec2(filename, data, result_path)
-else
-  abort('No valid platform. Valid platforms is ec2.')
+      run_command("az login",
+                  "--service-principal -u #{application_id} -p #{application_key} " \
+                  "--tenant #{job_data['subscription']} --debug",
+                  "Logging in as OBS app")
+
+      run_command("az storage container create",
+                  "-n #{job_data['container']} --account-name #{job_data['storage_account']} --debug",
+                  "Creating container at '#{job_data['storage_account']}/#{job_data['container']}'")
+
+      run_command("az storage blob upload",
+                  "--container-name #{job_data['container']} --account-name #{job_data['storage_account']} -f #{uncompressed_file} " \
+                  "-n #{remote_file_name} --debug",
+                  "Uploading image file '#{uncompressed_file}' to a blob")
+
+      source_url = uploaded_image_url(job_data['storage_account'], job_data['container'], remote_file_name)
+      result = run_command("az image create",
+                           "--resource-group #{job_data['resource_group']} --name #{image_name} " \
+                           "--source #{source_url} --os-type Linux --debug",
+                           "Creating image '#{image_name}' out of the blob '#{remote_file_name}'")
+      STDOUT.write("#{JSON.parse(result).inspect}\n")
+
+      run_command("az storage blob delete",
+                  "--container-name #{job_data['container']} --account-name #{job_data['storage_account']} -n #{remote_file_name} --debug",
+                  "Deleting")
+
+      run_command("az logout", "", "Logging out")
+    end
+
+    def self.decrypt(encrypted_data)
+      private_key = OpenSSL::PKey::RSA.new(File.read('/etc/cloudupload/secret.pem'))
+      encrypted_data.map { |encrypted_string| private_key.private_decrypt(Base64.decode64(encrypted_string)) }
+    end
+
+    def self.uncompress(backend_image_file)
+      file_path = File.join(Dir.pwd, backend_image_file)
+      compressed_file = "#{file_path}.xz"
+      uncompressed_file = "#{file_path}.vhd"
+      FileUtils.ln_s(backend_image_file, File.join(Dir.pwd, compressed_file))
+      run_command("xz", "-c -d #{compressed_file} > #{uncompressed_file}", "Uncompressing file '#{backend_image_file}'")
+      uncompressed_file
+    end
+
+    def self.calculate_image_name(image_filename)
+      image_filename.gsub(/\.xz|\.vhdfixed|\.vhd/, '')
+    end
+
+    def self.run_command(command, params, message, options = { logout_on_failure: true })
+      STDOUT.write("#{message}...")
+      out, err, status = Open3.capture3("#{command} #{params}")
```

Comment:
Wouldn't it be better to split it in ```Open3.capture3(command, params)```?

Reply:
Does that work actually? **params** are the parameters of the command, not the options for the `capture3` method

Reply:
> If opts is specified, it is sent to the command's standard input.

from the docs you mentioned. So yes, it seems to work.

Just tried: ```Open3.capture3("echo", "fooo", "bar")```

I think you just need to make sure to group parameters and values together in the array:

Like:
```
"az storage blob upload", ["--container-name #{job_data['container']}", "--account-name #{job_data['storage_account']}]
```

Reply:
But don't invest too much time in it

Reply:
I will change it, also taking into account what Marcus raised in his comment.